src/mrpro/operators/WaveletOp.py

```diff
@@ -0,0 +1,374 @@
+            current_shape = torch.as_tensor(domain_shape)
+
+            # if level is None, select the highest possible level.
+            # raise error/warnings if level is not possible or lead to boundary effects
+            verified_level = _check_level(domain_shape, wavelet_length, level)
+
+            if verified_level == 0:  # only a/aa/aaa component possible
+                self.coefficients_shape = [domain_shape]
+            else:
+                self.coefficients_shape = []
+                for _ in range(verified_level):
+                    # Add padding
+                    current_shape = (current_shape / 2).ceil() + wavelet_length // 2 - 1
+                    self.coefficients_shape.extend(
+                        [tuple(current_shape.to(dtype=torch.int64))] * self.n_wavelet_directions
+                    )
+
+                self.coefficients_shape = self.coefficients_shape[::-1]
+                self.coefficients_shape.insert(0, self.coefficients_shape[0])  # shape of a/aa/aaa term
+
+    def forward(self, x: torch.Tensor) -> tuple[torch.Tensor,]:
+        """Calculate wavelet coefficients from (image) data.
+
+        Parameters
+        ----------
+        x
+            (Image) data
+
+        Returns
+        -------
+            Wavelet coefficients stacked along one dimension
+
+        Raises
+        ------
+        ValueError
+            If the dimensions along which wavelets are to be calculated are not unique.
+        """
+        # normalize axes to allow negative indexing in input
+        dim = tuple(d % x.ndim for d in self._dim)
+        if len(dim) != len(set(dim)):
+            raise ValueError(f'Axis must be unique. Normalized axis are {dim}')
+
+        # move axes where wavelets are calculated to the end
+        x = torch.moveaxis(x, dim, list(range(-len(self._dim), 0)))
+
+        # the ptwt functions work only for real data, thus we handle complex inputs as an additional channel
+        x_real = torch.view_as_real(x).moveaxis(-1, 0) if x.is_complex() else x
+
+        if len(self._dim) == 1:
+            coeffs_1d = wavedec(x_real, self._wavelet_name, level=self._level, mode='zero', axis=-1)
+            coefficients_list = self._format_coeffs_1d(coeffs_1d)
+        elif len(self._dim) == 2:
+            coeffs_2d = wavedec2(x_real, self._wavelet_name, level=self._level, mode='zero', axes=(-2, -1))
+            coefficients_list = self._format_coeffs_2d(coeffs_2d)
+        elif len(self._dim) == 3:
+            coeffs_3d = wavedec3(x_real, self._wavelet_name, level=self._level, mode='zero', axes=(-3, -2, -1))
+            coefficients_list = self._format_coeffs_3d(coeffs_3d)
+        else:
+            raise ValueError(f'Wavelets are only available for 1D, 2D and 3D and not {self._dim}D')
+
+        # stack multi-resolution wavelets along single dimension
+        coefficients_stack = self._coeff_to_stacked_tensor(coefficients_list)
```

Comment:
Consider implementing logging at key operations within the wavelet transformation process to aid in debugging and performance monitoring. [medium]